### PR TITLE
New board pin-mapping system

### PR DESCRIPTION
**Pull request by chuckwagoncomputing**
This is the first step of #2144.
I have converted microrusefi to the new format, and confirmed that an identical rusefi_mre_f4.ini is generated.
This is a draft because:
1. This is the first non-HelloWorld Java I have written and I'm guessing it needs help
2. I need to convert the yamls for all the other boards

Update: This is ready to be reviewed/merged.

### CODE REVIEW

File: java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java [hunk 1]
```diff
@@ -325,6 +342,93 @@ public static void readPrependValues(VariableRegistry registry, String prependFi
...
+        Map<Integer, String> listSwitchInputs = new HashMap<>();
+        Map<String, Object> yamlData = yaml.load(new FileReader(yamlInputFile));
+        List<Map<String, Object>> data = (List<Map<String, Object>>) yamlData.get("pins");
+        if (data == null) {
+            SystemOut.println("Null yaml for " + yamlInputFile);
+        } else {
+            SystemOut.println(data);
+            Objects.requireNonNull(data, "data");
+            for (Map<String, Object> pin : data) {
+                if (pin.get("id").getClass().getName() == "java.util.ArrayList") {
```
Comment (mck1117) [inline java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java:359]: you can do something like `pin.get("id") instanceof ArrayList`

File: java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java [hunk 2]
```diff
@@ -325,6 +342,93 @@ public static void readPrependValues(VariableRegistry registry, String prependFi
...
+            registerPins(listOutputs, "output_pin_e_enum", "GPIO_UNASSIGNED", registry);
+            registerPins(listAnalogInputs, "adc_channel_e_enum", "EFI_ADC_NONE", registry);
+            registerPins(listEventInputs, "brain_input_pin_e_enum", "GPIO_UNASSIGNED", registry);
+            registerPins(listSwitchInputs, "switch_input_pin_e_enum", "GPIO_UNASSIGNED", registry);
+        }
+    }
+
+    private static int getMaxValue(Collection<Integer> values) {
+        int result = -1;
+        for (Integer v : values) {
```
Comment (mck1117) [inline java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java:379]: `return values.stream().max(Integer::compare).get()`

Comment (chuckwagoncomputing) [reply java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java]: I completely removed that function. Did I do it right?

Comment (chuckwagoncomputing) [reply java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java]: https://github.com/rusefi/rusefi/pull/2231/commits/e0e10241a6a3e3af3985149bfc54a9003b9beafe

Comment (mck1117) [reply java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java]: Looks good to me (assuming it compiles and works of course 🙂 )!

File: java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java [hunk 3]
```diff
@@ -325,6 +342,85 @@ public static void readPrependValues(VariableRegistry registry, String prependFi
...
     }
 
+    public static void processYaml(VariableRegistry registry, File yamlInputFile, ReaderState state) throws IOException {
+        Yaml yaml = new Yaml();
+        Map<Integer, String> listOutputs = new HashMap<>();
+        Map<Integer, String> listAnalogInputs = new HashMap<>();
+        Map<Integer, String> listEventInputs = new HashMap<>();
+        Map<Integer, String> listSwitchInputs = new HashMap<>();
+        Map<String, Object> yamlData = yaml.load(new FileReader(yamlInputFile));
+        List<Map<String, Object>> data = (List<Map<String, Object>>) yamlData.get("pins");
```
Comment (chuckwagoncomputing) [inline java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java:352]: Idea says: Unchecked cast: 'java.lang.Object' to 'java.util.List<java.util.Map<java.lang.String,java.lang.Object>>'

What does this mean, and how do I resolve it?

Comment (mck1117) [reply java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java:352]: Second answer is more useful: https://stackoverflow.com/questions/262367/type-safety-unchecked-cast

> Use @SuppressWarnings("unchecked") and hold your nose. Oh, and campaign for reified generics in Java :)

Just put `@SuppressWarnings("unchecked")` on the line before the function definition (345-ish).

**Comment by chuckwagoncomputing** [comment]
I take back the "identical rusefi_mre_f4.ini" being generated. I'm not sure what I did wrong when comparing, I must have not deleted the cache before running.

**Comment by rusefillc** [comment]
At the moment we have horrible practice of https://github.com/rusefi/rusefi/blob/master/java_tools/ConfigDefinition.jar being checked into repository

please either have a separate PR where we always compile https://github.com/rusefi/rusefi/blob/master/java_tools/ConfigDefinition.jar and remove https://github.com/rusefi/rusefi/blob/master/java_tools/ConfigDefinition.jar from GH or include latest https://github.com/rusefi/rusefi/blob/master/java_tools/ConfigDefinition.jar into this PR

**Comment by chuckwagoncomputing** [comment]
I will include the jar once this is ready to go.

**Comment by chuckwagoncomputing** [comment]
OK, got microrusefi generating properly.

**Comment by rusefillc** [comment]
maybe rebase not merge? at the moment https://github.com/rusefi/rusefi/pull/2231/files shows weird stuff

**Comment by rusefillc** [comment]
@chuckwagoncomputing PR description says "This is a draft because:

This is the first non-HelloWorld Java I have written and I'm guessing it needs help
I need to convert the yamls for all the other boards"

please update?

File: java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java [hunk 4]
```diff
@@ -325,6 +340,93 @@ public static void readPrependValues(VariableRegistry registry, String prependFi
...
+            List<Map<String, Object>> data = (List<Map<String, Object>>) yamlData.get("pins");
+            if (data == null) {
+                SystemOut.println("Null yaml for " + yamlFile);
+            } else {
+                SystemOut.println(data);
+                Objects.requireNonNull(data, "data");
+                for (Map<String, Object> pin : data) {
+                    if (pin.get("id") instanceof ArrayList) {
+                        for (int i = 0; i < ((ArrayList) pin.get("id")).size(); i++) {
+                            findMatchingEnum((String) ((ArrayList) pin.get("id")).get(i),
```
Comment (rusefillc) [inline java_tools/configuration_definition/src/com/rusefi/ConfigDefinition.java:361]: is that 5 levels of if/for deep? in an ideal world I would extract some smaller methods to break this super-method into bit simpler pieces?

**Comment by rusefillc** [comment]
this is _huge_ progress
WOW

**Comment by rusefillc** [comment]
https://github.com/rusefi/rusefi/issues/2234 created

**Pull request merged**